text/0000-template.md

```diff
@@ -0,0 +1,78 @@
+- Start Date: 2019-02-13
+- Relevant Teams: Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/445
+- Tracking: (leave this empty)
+
+# Deprecate `{{with}}`
+
+## Summary
+
+The `{{with}}` helper has always had slightly confusing conditional semantics, this was one of the motivators for [introducing](https://github.com/emberjs/rfcs/blob/master/text/0286-block-let-template-helper.md) the easier to understand `{{let}}` helper. Now that `{{let}}` exists, the remaining use case for using `{{with}}` is for its unique conditional semantics. These conditional semantics can be cleanly represented with a combination of `{{let}}` and `{{if}}` statements so ee should deprecate `{{with}}`.
+
+## Motivation
+
+The difference between `{{let}}` and `{{with}}` is with how they handle conditional arguments. The `{{let}}` helper's block content is always rendered, regardless of its parameters. In contrast, `{{with}}` only renders its main block when the first position parameter is truthy. For example:
+
+```hbs
+{{#with "Alex" as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "[Alex] is truthy".
+
+```hbs
+{{#with false as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "The first positional param was falsy".
+
+The conditional arguments behavior of `{{with}}` can easily be replicated using a combination of `{{let}}` and `{{if}}` in a way that's easily readable:
+
+```hbs
+{{#let "Alex" as |value|}}
+  {{#if value}}
+    {{value}} is truthy
+  {{else}}
+    The first positional param was falsy
+  {{/if}}
+{{/let}}
+```
+
+## Detailed design
+
+We'll create an AST transform in [`packages/ember-template-compiler`](https://github.com/emberjs/ember.js/tree/master/packages/ember-template-compiler) which will emit a deprecation warning for all uses of `{{with}}`. The deprecation warning will be something like:
+
+```
+Using `{{with}}` is deprecated, please use a combination of `{{let}}` and `{{if}}` instead.
+```
+
+This new deprecation warning should link to some documentation which clearly describes when you should use `{{let}}` and when you should use `{{let}}` and `{{if}}` in combination.
+
+## How we teach this
+
+We'll mentiton the deprecation in an Ember point release blog post.
+
+The deprecation message will link to good documentation which describe why `{{with}}` was deprecated and gives clear guidelines on how to migrate to using either `{{let}}` or `{{let}}` and `{{it}}` in combination.
```

Comment:
We should include the deprecation guide prose in this section directly. The deprecation guide entry will be linked to directly from the deprecation itself, and should show before/after examples for the various styles of `{{with` usage.